core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala

```diff
@@ -1865,93 +1928,29 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
-        case None =>
-          if (submissionRequest.entityType.getOrElse("") != rootEntityType) {
-            val whatYouGaveUs = if (submissionRequest.entityType.isDefined) s"an entity of type ${submissionRequest.entityType.get}" else "no entity"
-            DBIO.failed(new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.BadRequest, s"Method configuration expects an entity of type $rootEntityType, but you gave us $whatYouGaveUs.")))
-          } else {
-            withSingleEntityRec(submissionRequest.entityType.get, submissionRequest.entityName.get, workspaceContext, dataAccess)(rec => op(Some(rec)))
-          }
-        case Some(expression) =>
-          ExpressionEvaluator.withNewExpressionEvaluator(dataAccess, workspaceContext, submissionRequest.entityType.get, submissionRequest.entityName.get) { evaluator =>
-            evaluator.evalFinalEntity(workspaceContext, expression).asTry
-          } flatMap { //gotta close out the expression evaluator to wipe the EXPREVAL_TEMP table
-            case Failure(regret) => DBIO.failed(new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.BadRequest, regret)))
-            case Success(entityRecords) =>
-              if (entityRecords.isEmpty) {
-                DBIO.failed(new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.BadRequest, "No entities eligible for submission were found.")))
-              } else {
-                val eligibleEntities = entityRecords.filter(_.entityType == rootEntityType).toSeq
-                if (eligibleEntities.isEmpty)
-                  DBIO.failed(new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.BadRequest, s"The expression in your SubmissionRequest matched only entities of the wrong type. (Expected type ${rootEntityType}.)")))
-                else
-                  op(Some(eligibleEntities))
-              }
-          }
-      }
-    }
-  }
 
   /** Validates the workflow failure mode in the submission request. */
-  private def withWorkflowFailureMode(submissionRequest: SubmissionRequest)(op: Option[WorkflowFailureMode] => ReadWriteAction[PerRequestMessage]): ReadWriteAction[PerRequestMessage] = {
+  private def getWorkflowFailureMode(submissionRequest: SubmissionRequest): Future[Option[WorkflowFailureMode]] = {
     Try(submissionRequest.workflowFailureMode.map(WorkflowFailureModes.withName)) match {
-      case Success(failureMode) => op(failureMode)
-      case Failure(NonFatal(e)) => DBIO.failed(new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.BadRequest, e.getMessage)))
+      case Success(failureMode) => Future.successful(failureMode)
+      case Failure(NonFatal(e)) => Future.failed(new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.BadRequest, e.getMessage)))
     }
   }
 
-
-  private def withSubmissionParameters(workspaceName: WorkspaceName, submissionRequest: SubmissionRequest)
```

Comment:
this function was essentially rewritten to get rid of all the withs and break up the database interaction to make way for data repo and big query calls.